lp2p/identity.go

```diff
@@ -0,0 +1,80 @@
+package lp2p
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const keyFilename = "p2p.key"
+
+type identityInfo struct {
+	Key []byte
+	ID  peer.ID // this is needed only to simplify integration with some testing tools
+}
+
+func genIdentity() (crypto.PrivKey, error) {
+	pk, _, err := crypto.GenerateEd25519Key(rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ed25519 identity: %w", err)
+	}
+	return pk, nil
+}
+
+func identityInfoFromDir(dir string) (*identityInfo, error) {
+	path := filepath.Join(dir, keyFilename)
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %s: %w", path, err)
+	}
+	var info identityInfo
+	err = json.Unmarshal(data, &info)
+	if err != nil {
+		return nil, fmt.Errorf("faile to unmarshal file content from %s into %+v: %w", path, info, err)
+	}
+	return &info, nil
+}
+
+func ensureIdentity(dir string) (crypto.PrivKey, error) {
+	// TODO add crc check
+	if err := os.MkdirAll(dir, 0o755); err != nil {
+		return nil, fmt.Errorf("can't ensure that directory %s exist: %w", dir, err)
+	}
+	info, err := identityInfoFromDir(dir)
+	if err == nil {
+		return crypto.UnmarshalPrivateKey(info.Key)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		key, err := genIdentity()
+		if err != nil {
+			return nil, err
+		}
+		id, err := peer.IDFromPrivateKey(key)
+		if err != nil {
+			panic("generated key is malformed")
+		}
+		raw, err := crypto.MarshalPrivateKey(key)
+		if err != nil {
+			panic("generated key can't be marshaled to bytes")
+		}
+		data, err := json.Marshal(identityInfo{
+			Key: raw,
+			ID:  id,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if err := ioutil.WriteFile(filepath.Join(dir, keyFilename), data, 0o644); err != nil {
+			return nil, fmt.Errorf("failed to write identity data: %w", err)
+		}
+		return key, nil
+	}
+	return nil, fmt.Errorf("failed to read key from disk: %w", err)
```

Comment:
If this returns an error, somewhere above the call stack we'll get a log like `Failed to start the node. See logs for details err=failed to initialize p2p host: failed to read key from disk: failed to read file %s: %w`, so words `failed to` will be duplicated a lot of times and the log will be unreadable. I suggest removing verbs (`failed to`/`cannot`/`unable to` etc) from all wrapped errors and having them only where we log an error.

Reply:
doesn't make sense to me, they shouldn't be wrapped then. how would you reword this error so that it makes sense?

Reply:
In my opinion, `Failed to start the node. See logs for details err=initialize p2p host: read key from disk: read file %s: %w` would be more readable, so I'd change this to
```suggestion
	return nil, fmt.Errorf("read key from disk: %w", err)
```

Reply:
done

Reply:
thanks